governance/validatorSet.go

```diff
@@ -0,0 +1,185 @@
+	client, err := ethclient.Dial(gethIpc)
+	if err != nil {
+		return common.Address{}, err
+	}
+	defer client.Close()
+	
+	ctx := context.Background()
+	cfg := txclient.NewTxDefaultConfig(deployContractGasLimit)
+
+	txOps, err := txclient.GenerateTxOpts(ctx, client, txAuth, cfg)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	addr, _, _, err := bindings.DeployValidatorSet(txOps, client)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	return addr, nil
+}
+
+func (v ValidatorSet) AddValidator(txAuth authy.Auth, pubKey string, address common.Address) (error) {
+	client, err := ethclient.Dial(v.gethIpc)
+	if err != nil {
+		return err
+	}
+
+	defer client.Close()
+
+	contractInstance, err := bindings.NewValidatorSet(v.contractAddress, client)
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	cfg := txclient.NewTxDefaultConfig(deployContractGasLimit)
+	txOps, err := txclient.GenerateTxOpts(ctx, client, txAuth, cfg)
+	if err != nil {
+		return err
+	}
+
+	tx, err := contractInstance.AddValidator(txOps, pubKey, address)
+	if err != nil {
+		return err
+	}
+	
+	_, err = txclient.FetchReceipt(client, tx, cfg)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (v ValidatorSet) RemoveValidator(txAuth authy.Auth, pubKey string, address common.Address) (error) {
+	client, err := ethclient.Dial(v.gethIpc)
+	if err != nil {
+		return err
+	}
+
+	defer client.Close()
+
+	contractInstance, err := bindings.NewValidatorSet(v.contractAddress, client)
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	cfg := txclient.NewTxDefaultConfig(deployContractGasLimit)
+	txOps, err := txclient.GenerateTxOpts(ctx, client, txAuth, cfg)
+	if err != nil {
+		return err
+	}
+
+	tx, err := contractInstance.RemoveValidator(txOps, pubKey, address)
+	if err != nil {
+		return err
+	}
+	
+	_, err = txclient.FetchReceipt(client, tx, cfg)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (v ValidatorSet) FetchPubKeySet(address common.Address) ([]string, error) {
+	client, err := ethclient.Dial(v.gethIpc)
+	if err != nil {
+		return nil, err
+	}
+
+	defer client.Close()
+
+	contractInstance, err := bindings.NewValidatorSet(v.contractAddress, client)
+	if err != nil {
+		return nil, err
+	}
+
+	validatorSetSizeBN, err := contractInstance.ValidatorSetSize(&bind.CallOpts{})
+	if err != nil {
+		return nil, err
+	}
+	validatorSetSize :=int(validatorSetSizeBN.Int64())
+	var validatorSet []string
+	for i := 0; i < validatorSetSize; i++ {
+		validatorPubKey, err := contractInstance.ValidatorPubKey(&bind.CallOpts{}, big.NewInt(int64(i)))
+		if err != nil {
+			return nil, err
+		}
+		validatorSet = append(validatorSet, validatorPubKey)
+	}
+
+	return validatorSet, nil
+}
+
+func (v ValidatorSet) ValidatorAddress(pubKey string) (common.Address, error) {
+	if v.contractAddress.String() == common.HexToAddress("0x0").String() {
+		return common.Address{}, nil
+	}
+
+	client, err := ethclient.Dial(v.gethIpc)
```

Comment:
[critical] This function is being executed twice on every block. I think we should pass the eth client as an argument in this case to avoid opening and closing connection all the time. 

I did it like this for some Leth functions that I didn't expect to happen often for design simplicity but this will be running all the time.

Optimally we could even cache in memory the result of the operation BUT you would have to remember to update the relation between ValidatorAddress -> EthereumAccount every time a validator is removed/added OR the version of smart contract changed. I leave that to your judgement but at least optimizing the opening and closing connection should be done imo.

Reply:
I do understand and share your concerns. As far as I was capable to validate from this solution, there is not any significative impact over the usage of a socket file to perform this read actions over the blockchain using the RPC protocol. 

The usage of a cache would add a large overhead to the code for very little improvements. Let's discuss this offline try to make numbers around those solutions.

Reply:
Oki, after an offline we decide to use a single connector which remains open during the entire execution